File: Data Structures & Algorithms/course-schedule/submission-1.cpp
class Solution {
private:
    std::vector<std::vector<int>> create_adj_list(int num_courses, const std::vector<std::vector<int>>& pre_reqs) {
        std::vector<std::vector<int>> adj_list(num_courses);

        for(const auto& pair : pre_reqs) {
            int course = pair[0];
            int pre_req = pair[1];

            adj_list[course].push_back(pre_req);
        }
        return adj_list;
    }

    bool has_cycle(int course, std::vector<int>& state, const std::vector<std::vector<int>>& adj_list) {
        if(state[course] == -1) {
            return true;
        }
        if (state[course] == 1) {
            return false;
        }

        state[course] = -1;

        for (int pre_req : adj_list[course]) {
            if (has_cycle(pre_req, state, adj_list)) {
                return true;
            }
        }

        state[course] = 1;
        return false;

    }

public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        std::vector<std::vector<int>> adj_list = create_adj_list(numCourses, prerequisites);
        std::vector<int> state(numCourses, 0);

        for (int course {0}; course < numCourses; ++course) {
            if (has_cycle(course, state, adj_list)) {
                return false;
            }
        }
        return true;
    }
};

File: Data Structures & Algorithms/course-schedule-ii/submission-1.py
class Solution:
    def findOrder(self, num_courses: int, prerequisites: List[List[int]]) -> List[int]:
        result = []

        adj_list = [[] for _ in range(num_courses)]
        for c1, c2 in prerequisites:
            adj_list[c1].append(c2)


        state = [0] * num_courses

        def has_cycle(course):
            if state[course] == 1:
                return False
            
            if state[course] == -1:
                return True

            state[course] = -1

            for neighbors in adj_list[course]:
                if has_cycle(neighbors):
                    return True
            
            state[course] = 1
            result.append(course)
            return False


        for course in range(num_courses):
            if has_cycle(course):
                return []

        return result
        